sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,475 @@
+rectangular). This can be achieved by using a pair or triple instead of a
+single float wherever a parameter value is specified. In case of a pair,
+the first value specifies the lower boundary and the second value the upper
+boundary. In case of a triple, the middle value specified the initial value
+of the parameter during hyperparameter-optimization.
+"""
+
+# Author: Jan Hendrik Metzen <jhm@informatik.uni-bremen.de>
+# Licence: BSD 3 clause
+
+# Note: this module is strongly inspired by the kernel module of the george
+#       package.
+
+import numpy as np
+from scipy.spatial.distance import pdist, cdist, squareform
+
+
+class Kernel(object):
+    """ Base class for all kernels."""
+
+    def _parse_param_space(self, param_space):
+        if not hasattr(param_space, "__iter__"):  # fixed hyperparameter
+            self.params = np.array([float(param_space)])
+            self.has_bounds = False
+            return
+        param_space = np.atleast_2d(param_space)
+        if param_space.shape[1] == 1:  # fixed hyperparameter
+            self.params = param_space[:, 0]
+            self.has_bounds = False
+        elif param_space.shape[1] == 2:  # lower+upper bound for hyperparameter
+            self.bounds = param_space
+            self.has_bounds = True
+            # Use geometric mean of upper and lower boundary as initial
+            # hyperparameter value
+            assert not np.any(self.l_bound == None)  # XXX: enforce element-wise comparison to None
+            assert not np.any(self.u_bound == None)
+            self.params = np.array([np.sqrt(self.l_bound * self.u_bound)])
+        elif param_space.shape[1] == 3:  # lower bound, initial value, upper bound
+            self.params = param_space[:, 1]
+            self.bounds = param_space[:, [0, 2]]
+            self.has_bounds = True
+        else:
+            raise Exception()
+
+    @property
+    def n_params(self):
+        return self.params.shape[0]
+
+    @property
+    def bounds(self):
+        return np.vstack((self.l_bound, self.u_bound)).T
+
+    @bounds.setter
+    def bounds(self, bounds):
+        bounds = bounds.reshape(-1, 2)
+        self.l_bound = bounds[:, 0]
+        self.u_bound = bounds[:, 1]
+
+    def __add__(self, b):
+        if not isinstance(b, Kernel):
+            return Sum(self, ConstantKernel(b))
+        return Sum(self, b)
+
+    def __radd__(self, b):
+        if not isinstance(b, Kernel):
+            return Sum(ConstantKernel(b), self)
+        return Sum(b, self)
+
+    def __mul__(self, b):
+        if not isinstance(b, Kernel):
+            return Product(self, ConstantKernel(b))
+        return Product(self, b)
+
+    def __rmul__(self, b):
+        if not isinstance(b, Kernel):
+            return Product(ConstantKernel(b), self)
+        return Product(b, self)
+
+    def __repr__(self):
+        return "{0}({1})".format(self.__class__.__name__,
+                                 ", ".join(map("{0}".format, self.params)))
+
```

Comment:
Can you add `def auto(self, X): raise NotImplementedError('Use subclasses of Kernel, for example ...')`, same for `def cross`.

While we are at it: Couldn't `__call__ == cross`? You could just treat the kernel as a callable for its most general usage.